### PR TITLE
Hugging face model deployment with Ray Serve

### PR DESCRIPTION
**Pull request by pthombre**
# What does this PR do ?
Set up framework for using Ray Serve for model deployments. Currently this only supports HuggingFace models

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [X ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Review by oyilmaz-nvidia — COMMENTED** [review]
So, overall, this looks good, and thanks for following the existing design patter for deploying models with PyTriton.

The only question I have in my mind is that is there any common functions that you think we should force users to implement to deploy anything with Ray? For instance, you can ask users to implement completion function as an example, or anything you think it should be a must.

File: scripts/deploy/nlp/deploy_ray_hf.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+import logging
+import signal
+import sys
+from typing import Optional
+
+from nemo.deploy.deploy_ray import DeployRay
+from nemo.deploy.nlp.hf_deployable_ray import HFRayDeployable
+
+LOGGER = logging.getLogger("NeMo")
+
```
Comment (oyilmaz-nvidia) [inline scripts/deploy/nlp/deploy_ray_hf.py:25]: Do you think it makes sense to merge the ray and pytriton based deployment in one script and make the deploy option an argument?

Same thing can be done for the query as well.

File: nemo/deploy/deploy_ray.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            # Try to connect to existing Ray cluster
+            ray.init(address=address, ignore_reinit_error=ignore_reinit_error, runtime_env=runtime_env)
+        except ConnectionError:
+            # If no cluster exists, start a local one
+            LOGGER.info("No existing Ray cluster found. Starting a local Ray cluster...")
+            ray.init(
+                num_cpus=num_cpus,
+                num_gpus=num_gpus,
+                include_dashboard=include_dashboard,
+                ignore_reinit_error=ignore_reinit_error,
```
Comment (athitten) [inline nemo/deploy/deploy_ray.py:117]: Just curious: it sounds like we want to have `ignore_reinit_error` as False by default. Is it not the case ?

File: nemo/deploy/nlp/hf_deployable.py [hunk 3]
```diff
@@ -255,65 +349,58 @@ def triton_infer_fn(self, **inputs: np.ndarray):
...
-                    for r in output["logits"]:
-                        lp = torch.tensor(r).cpu().detach().numpy()
-                        if len(lp) == 0:
-                            output_logits.append([0])
-                        else:
-                            output_logits.append(lp)
-                    output_infer["logits"] = np.array(output_logits).transpose(1, 0, 2)
-            else:
-                output_infer = {"sentences": cast_output(output, np.bytes_)}
+    def ray_infer_fn(self, inputs: Dict[Any, Any]):
```
Comment (athitten) [inline nemo/deploy/nlp/hf_deployable.py:367]: @pthombre not sure if this out of scope for the PR but can we unify `ray_Infer_fn` and `triton_infer_fn` ? They share the same logic.

File: scripts/deploy/nlp/deploy_ray_hf.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    parser.add_argument(
+        "--num_gpus_per_replica",
+        type=float,
+        default=1,
+        help="Number of GPUs per model replica",
+    )
+    parser.add_argument(
+        "--num_cpus_per_replica",
+        type=float,
+        default=8,
```
Comment (athitten) [inline scripts/deploy/nlp/deploy_ray_hf.py:113]: Should we make this as 1 ?

**Comment by athitten** [comment]
@pthombre sorry for the delay. Thank you for this work, this looks great! Left some comments. Few follow ups:

1) Do you see any perf difference between using Pytriton v/s Ray for MegatronLLMDeployable? Its probably unlikely, but don't want to exclude it also.
2) Also when we use multi instance or replicas where in the code is output from all replicas get collated ? Or does this not exist because each replica works as an independent module and sends the outputs asynchronously ?
3) If its not difficult, can we add some perf numbers on some small model by doing multi instance query v/s single instance ?
